### PR TITLE
Full Server Impl Random Events

### LINKED ISSUE #59: Random Events Framework
Create the initial framework for Random Events on the server.

Random Events should be created and distributed by a background process that executes regularly, selecting a number of random players from those who are in the Online Users cache and sending them an event.

Random events should happen to a given player approximately once every two hours of online play. Ensure that players don't get random events too often (if the process cycles every 45seconds and you get an event each time, it's gonna feel wack). This may mean caching players who have received an event in a blacklist with a TTL for when they become eligible again.

Random events can be trivial

> You find $20 on the ground

> A jack-o-lantern says Hello!

Or significant

> You find a rare gem

And they can be negative

> The Sharewolf attacks, taking $100 from you.

Events should be configurable, and should be stored as a JSON serialized class in the DB.

Processing logic for an Event is driven by the information in the config, so the `RandomEventConfig` class and row will need to hold a lot of well defined information:

```
public class RandomEventConfig
{
    DateTime? EventIsValidStart;
    DateTime? EventIsValidEnd;
    String ImageUrl;
    String DisplayMessage;
    List<ItemTypeConfigs> SpecificTransferableItems;
    Int? ItemRarityThresholdMin;
    Int? ItemRarityThresholdMax;
    Int? CashDeltaMax;
    Int? CashDeltaMin;
    Bool? TGiveFTakeNullNeither;
}
```

Some of this may make more sense to split out of the class and keep as raw SQL. Implementer's Choice.

Random Events will be transmitted to online players via the SignalR IHubContext that is maintained for all active players, passing a RandomEvent object to the UI that will be used to create a notification and display the outcome to the player. The RandomEvent will be passed over SignalR but interpreted just like an API Response object. (SignalR allows the server to push to the User without waiting on an initiating request)

Just like NPCs, RandomEvents may:

-  Have images and text that need to be displayed in the UI

> "A passerby says: Hello! " 
> image

- cause effects to the User that need to be explained
 
>    "The Sharewolf appears and steals a hundred of your dollars!"
> image
> 

- need to trigger UI effects (confetti CSS-animation if an item is awarded)

So the response must be structured in an easily templated-way
```
List<InventoryItemDetails> GivenItems;
List<InventoryItemDetails> TakenItems;
String DisplayMessage;
String DisplayImageURL;
Int GivenCash;
Int TakenAwayCash;
```

We are not passing HTML for the UI to inject, the UI will parse out a well-structured Response object to create a response UI element.

### PR DESCRIPTION
**Pull request by Jeremy-Barnes**
An IHostedService with a routine that executes every 90 seconds, selecting a set of users such that that over a 2 hour period, every online user should get a random event.

Users that have received an event recently are cached in Redis so they don't get too many events.

Events are stored in the DB as serialized JSONB representations of the RandomEventDefinition object, which contains the actual rules for generation.

Users are notified via SignalR that they have received an event.

Fixes #59

### CODE REVIEW

File: CritterServer/Domains/EventDomain.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                                //save them to redis, but don't wait up
+                                _ = randomEventRepo.SaveRandomEventReceipt(winner);
+                            }
+                        }
+                        else
+                        {
+                            duplicateWinnerCount++;
+                        }
+                    }
+                    luckyWinnerUsernames = await userRepo.RetrieveOnlineUsernames(randomSamplingCount: winnerCount);
```
Comment (acmmlu) [inline CritterServer/Domains/EventDomain.cs:73]: Don't we only need to retrieve as many new winners as there were duplicates?

File: CritterServer/Domains/EventDomain.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+using Serilog;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CritterServer.Domains
+{
+    public class RandomEventService : IHostedService
```
Comment (acmmlu) [inline CritterServer/Domains/EventDomain.cs:19]: seems worthy of its own file

File: CritterServer/Domains/EventDomain.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            await CreateRandomEvent(user, activeDev);
+        }
+
+        public async Task<bool> CreateRandomEvent(User recipient, User activeDev)
+        { 
+            RandomEventNotification notify = new RandomEventNotification();
+
+            var eventDefs = await RandomEvents.GetEventDefinitions();
+            if (eventDefs.Count() == 0) return false;
+            var selectedDefId = random.Next(0, eventDefs.Count());
```
Comment (acmmlu) [inline CritterServer/Domains/EventDomain.cs:122]: Configurable weights for event types? (this might be a feature suggestion lol)

**Review by acmmlu — APPROVED** [review]

File: CritterServer/Domains/EventDomain.cs (hunk 3 above)
Comment (Jeremy-Barnes) [reply CritterServer/Domains/EventDomain.cs:122]: Big agree, I almost did it out of the box but it was a bit more of a pain in the ass than I felt was strictly necessary for the moment.

**Pull request merged**